platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/schemaBuilder/schemaBuildingUtils.kt

```diff
@@ -68,11 +68,27 @@ val KCallable<*>.annotationsWithGetters: List<Annotation>
     get() = this.annotations + if (this is KProperty) this.getter.annotations else emptyList()
 
 
-fun KType.toDataTypeRefOrError() =
-    toDataTypeRef() ?: error("failed to convert type $this to data type")
+fun KCallable<*>.returnTypeToRefOrError(receiver: KClass<*>?) =
+    returnTypeToRefOrError(receiver) { this.returnType }
+
+
+fun KCallable<*>.returnTypeToRefOrError(receiver: KClass<*>?, typeMapping: (KCallable<*>) -> KType) =
+    typeMapping(this).toDataTypeRef() ?: interpretationFailure("Conversion to data types failed for return type of ${format(receiver, this)}: ${typeMapping(this)}")
+
+
+fun KParameter.parameterTypeToRefOrError(receiver: KClass<*>?, function: KFunction<*>) =
+    parameterTypeToRefOrError(receiver, function) { this.type }
+
+
+fun KParameter.parameterTypeToRefOrError(receiver: KClass<*>?, function: KFunction<*>, typeMapping: (KParameter) -> KType) =
+    typeMapping(this).toDataTypeRef() ?: interpretationFailure("Conversion to data types failed for parameter type of function ${format(receiver, function)}: ${typeMapping(this)}")
```

Comment:
It might be nice to also say which parameter it is, especially since `format` does not list parameters.